Remove ShardManagerMonitor and other gateway cleanup

kangalio opened the pull request:
The bulk of this PR is removing ShardManagerMonitor. It was just a background task that received ShardManagerMessage's and called the respective ShardManager function. Now, you can just call the respective ShardManager function directly.

The commits about removing ShardManagerMonitor are marked with `[SMM]`. The other commits do other miscellaneous related clean-up. See commit titles and descriptions for more details.

Largely untested

kangalio committed:
Remove redundant ClientEvent and dispatch_client

dispatch_event was a function that was only ever called inside ShardRunner::run with a ClientEvent argument, which itself is just a single-variant enum that only ever contains a ShardStageUpdateEvent.

So, both ClientEvent and dispatch_event are confusing indirections without no clear purpose and I factored them out by inlining the implementation of dispatch_event into ShardRunner::run

kangalio committed:
Remove InterMessage

It was a single-variant enum that only ever contained ShardClientMessage. Using ShardClientMessage directly is less confusing

kangalio committed:
Implement ShardMessenger::send_to_shard TODO

kangalio committed:
[SMM] Reroute shard manager communication

From ShardManagerMonitor to ShardManager

First step in removing ShardManagerMonitor

kangalio committed:
[SMM] Inline Restart and ShardUpdate code into ShardManager

The message channel for ShardRunner was actually passing around Shard**Client**Message's instead of ShardRunnerMessage's (ShardClientMessage is an enum over ShardRunnerMessage and ShardManagerMessage). That's because ShardRunner hijacked two of ShardManagerMessage's enum variants for its own purpose (Restart and Shutdown). I moved those two to ShardRunnerMessage. And then replaced ShardClientMessage with ShardRunnerMessage for the ShardRunner channel communication.

After that, the only remaining uses of the ShardManagerMessage's enum variants were for communication with ShardManager (bliss!). So I removed the Restart and ShardUpdate variants in favor of executing their logic in ShardManager::{restart, shard_update} directly.

kangalio committed:
[SMM] ShardRunnerMessage::Shutdown{Initiated, Finished} were unused

kangalio committed:
[SMM] Inline ShutdownFinished code into ShardManager

kangalio committed:
[SMM] Replace ShardManageMonitor return value with channel

Basically, the only remaining purpose of ShardManagerMonitor was to let Client::start_connection() call it and get the return value back, once ShardManagerMessage::{ShutdownInitiated, ShardInvalidAuthentication, ShardInvalidGatewayIntents, ShardDisallowedGatewayIntents} is received. If we want to inline that code into ShardManager, we need to have a way to get the return value from a ShardManager method call into Client::start_connection. Obvious solution: a channel! So, this commit finally deletes ShardManagerMonitor and replaced the channel to it with a channel for the return value. And Client now stores that channel instead of ShardManagerMonitor and queries it inside start_connection

kangalio committed:
Remove Shard{Manager, Client}Message (unused now)

mkrasnitski commented:
Will review this later today. Just quickly noticed that the docs CI job is failing.

kangalio committed:
Fix docs

mkrasnitski commented on src/client/bridge/gateway/shard_manager.rs:
```diff
@@ -117,30 +111,43 @@ pub struct ShardManager {
             framework: opt.framework,
             last_start: None,
-            manager_tx: thread_tx.clone(),
+            manager: manager.clone(),
```
```suggestion
            manager: Arc::clone(&manager),
```

To indicate that this is just a refcount increment.

mkrasnitski commented on src/client/bridge/gateway/mod.rs:
```diff
@@ -43,7 +43,7 @@
 
 mod shard_manager;
-mod shard_manager_monitor;
+// mod shard_manager_monitor;
```
```suggestion
```

mkrasnitski commented on src/client/bridge/gateway/shard_queuer.rs:
```diff
@@ -187,7 +185,7 @@ impl ShardQueuer {
             #[cfg(feature = "framework")]
             framework: self.framework.get().map(Arc::clone),
-            manager_tx: self.manager_tx.clone(),
+            manager: self.manager.clone(),
```
```suggestion
            manager: Arc::clone(&self.manager),
```
Ditto.

mkrasnitski reviewed (changes requested):
I'd say this looks good, aside from a few tiny nits. Untangling the mess that is the gateway code is definitely a good thing (I say this as someone relatively unfamiliar with gateway internals).

Could you provide a sentence or two describing the breaking changes in this PR, for changelog purposes?

kangalio committed:
Apply suggestions from code review

Co-authored-by: Michael Krasnitski <42564254+mkrasnitski@users.noreply.github.com>

kangalio commented:
```
- Removed `InterMessage`, `ShardClientMessage`, `ShardManagerMessage`, `ShardManagerMonitor`, `ShardManagerError`
- Changed `ShardManager::new` return type from `(_, ShardManagerMonitor)` to `(_, UnboundedReceiver<Result<(), GatewayError>)`
  - Instead of calling `ShardManagerMonitor.run()`, you just wait for the return value to arrive via the `UnboundedReceiver`
- Changed `ShardMessenger::new` parameter type from `Sender<InterMessage>` to `Sender<ShardRunnerMessage>`
  - All shard-relevant actions have been centralized into `ShardRunnerMessage`
- Changed `ShardMessenger::send_to_shard` return type from `Result<_, _>` to `()`
  - Errors are now logged in the function directly
- Changed `ShardQueuer`'s and `ShardRunnerOptions`'s field `manage_tx: Sender<ShardManagerMessage>` field to `manager: Arc<Mutex<ShardManager>>`
  - Instead of invoking actions on `ShardManager` by sending messages through a channel, you now invoke methods on `ShardManager` directly
```

Since you said "a sentence or two":
```
`ShardManagerMonitor` has been removed in favor of interacting with `ShardManager` directly.
```

kangalio commented:
Here is some additional context about the removed types:

Removed:
- `ShardClientMessage`
    - used to send commands to a ShardRunner. But we already had a type for that: `ShardRunnerMessage`. `ShardClientMessage` existed solely because some shard runner commands ended up in the `ShardManagerMessage` enum for mysterious reasons, so `ShardClientMessage` was an enum over either `ShardRunnerMessage` or `ShardManagerMessage`. This PR moves all shard runner commands into `ShardRunnerMessage`, so you should use that type instead now.
- `InterMessage`
    - an enum that only ever contained `ShardClientMessage`. Use that type directly; or rather it's new replacement `ShardRunnerMessage`.
- `ShardManagerMonitor`
    - received `ShardManagerMessage`'s and propagated them to the actual shard manager. This redundancy has been removed and you can instead call methods on `ShardManager` directly now. All places that previously stored a `Sender<ShardManagerMessage`> now have a `Arc<Mutex<ShardManager>>`.
- `ShardManagerMessage`
    - see above
- `ShardManagerError`
    - duplicated parts of `GatewayError` and was removed in favor of `GatewayError`.

Thanks to @tazz4843 for pointing out the lacking explanation
